### PR TITLE
Enhancement cloudbackup v2 phoenix api support

### PR DESCRIPTION
**Pull request by BenjamenMeyer**
- Adds functionality for the up-coming v2.0 RCBU API
- Adds a command-line shell environment to interact with the API

### CODE REVIEW

**Comment by BenjamenMeyer** [comment]
@jc7998 @BryanSD I updated this based on the Agent CI/CD functionality, but the API doesn't seem to like it as the BackupConfigurationId is not specified - "BackupConfigurationId: null" in the JSON submitted. Further, I'm not sure why the CI/CD functionality isn't calling the "to_creation_dict" that strips several fields (BackupConfigurationId included) for creating new backups.

Could you two review and see what I'm missing?

File: cloudbackup/cmd/shell.py [hunk 1]
```diff
@@ -0,0 +1,1009 @@
...
+                'minute': None,
+                'amOrPm': None,
+                'timeZone': None
+            },
+            'interval': None
+        }
+
+        def promptBaseRate():
+            base_rate_menu = [
+                { 'index': 0, 'text': 'Manual', 'type': 'frequency' },
```
Comment (jc7998) [inline cloudbackup/cmd/shell.py]: Should be Manually

**Comment by jc7998** [comment]
@BenjamenMeyer 
You are right that the Integration Tests do not call to_creation_dict() .
In the Tests repo, the BackupConfigurationId entry is not present in  CreateBackupConfigurationTemplate() and BackupConfiguration is never used to store the Config Id; the tests retrieve it from API later by looking into all configurations available for the agent and then matches the name of the configuration.

Since the SDK is not calling to_creation_dict() either and its CreateBackupConfigurationTemplate() has an entry for BackupConfigurationId with a value of None, I think that when it sends the request to create the configuration it includes that value and API returns that incorrect error message.

Re: BackupConfigurationScheduled, I assume that the Integration Tests send this value to the API with a value of None and that the API simply ignores it. No json schema validation?

**Comment by jc7998** [comment]
@BenjamenMeyer 
You can find the request that the sdk submits here:
https://gist.github.com/jc7998/448f9583602772e504f8

I got the request to succeed by just removing the BackupConfigurationId entry. I left the BackupConfigurationScheduled as it was.

File: cloudbackup/cmd/shell.py [hunk 2]
```diff
@@ -0,0 +1,1009 @@
...
+        backup_config.StartTimeMinute = config_data['schedule']['start-time']['minute']
+        backup_config.StartTimeAmPm = config_data['schedule']['start-time']['am-pm']
+        if not config_data['schedule']['day-of-week'] is None:
+            backup_config.DayOfWeekId = DayOfWeekMapping[
+                config_data['schedule']['day-of-week']
+            ]
+        else:
+            backup_config.DayOfWeekId = None
+        backup_config.HourInterval = config_data['schedule']['hourly-interval']
+        backup_config.TimeZoneId = config_data['schedule']['start-time']['timezone']
```
Comment (jc7998) [inline cloudbackup/cmd/shell.py]: I think this is a problem.
If you choose an hourly backup (or manual one), it is overriding the default value with None.

**Comment by brett-johnson** [comment]
@jc7998 Thanks so much for the code review!

**Comment by BryanSD** [comment]
Agreed, I think the API code is just getting hung up because it's expecting an integer value for the `BackupConfigurationId` and you're providing `null`. In reality you don't need to provide that property at all - this is only a problem because the "create" and "get" API calls both share the same view model on the backend, even though only the "get" view model populates and uses `BackupConfigurationId`.

File: cloudbackup/cmd/shell.py (hunk 2 above)
Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: IIRC, the *original* default value *was* ```None```. Just checked and that seems to no longer be the case, so good catch. Thanks!
:money_with_wings:

File: cloudbackup/cmd/shell.py (hunk 1 above)
Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: Thanks. Fixed

**Comment by BenjamenMeyer** [comment]
@jc7998 please review again. I've submitted and had accepted two manual backup configurations; I haven't tried others yet. If all looks good, then I'll merge it. :smile_cat:

File: cloudbackup/client/agents.py [hunk 3]
```diff
@@ -351,49 +432,69 @@ def IsEncrypted(self):
...
 
     @property
     def Status(self):
         """
         Agent Status
         """
-        return self._details['Status']
+        if self.version == 1:
+            return self._details['Status']
+        # TODO: API v2 does not return status like this
```
Comment (jc7998) [inline cloudbackup/client/agents.py]: For reference, phoenix returns the status of the agent in a different endpoint:
http://docs.cloudbackupapi.apiary.io/#reference/agents/v2agentsidstatus/get-an-agent's-status

You'll have to use Chrome to follow that link. Firefox does not follow the url correctly due to the single quote.

File: cloudbackup/client/agents.py [hunk 4]
```diff
@@ -351,49 +432,69 @@ def IsEncrypted(self):
...
 
     @property
     def TimeOfLastSuccessfulBackup(self):
         """
         When was the agent last succcessful with its backup?
         """
-        return self._details['TimeOfLastSuccessfulBackup']
+        if self.version == 1:
+            return self._details['TimeOfLastSuccessfulBackup']
+        # TODO: API v2 does not return info like this
```
Comment (jc7998) [inline cloudbackup/client/agents.py]: v2 provides this info in a round about way.
You can find the last backup taken for a given configuration here:
http://docs.cloudbackupapi.apiary.io/#reference/configurations/v2configurationsid/get-details-about-a-configuration

and find the last_completed backup (snapshot) id.
With that, you can find ended_time here:
http://docs.cloudbackupapi.apiary.io/#reference/backups/v2backupsid/get-details-about-a-backup

File: cloudbackup/cmd/shell.py [hunk 5]
```diff
@@ -0,0 +1,1068 @@
...
+            config_data['schedule']['start-time']['hour'] = (
+                config_data['schedule']['start-time']['hour']
+                - 1
+            )
+            if config_data['schedule']['start-time']['am-pm'] == 'PM':
+                # Convert from AM to PM
+                config_data['schedule']['start-time']['hour'] = (
+                    config_data['schedule']['start-time']['hour']
+                    + 12
+                )
```
Comment (jc7998) [inline cloudbackup/cmd/shell.py]: I think you need to handle the scenario when the user selects 12 AM.
(haven't tested it, just looking into the code)

File: cloudbackup/client/agents.py (hunk 3 above)
Comment (BenjamenMeyer) [inline cloudbackup/client/agents.py]: http://docs.cloudbackupapi.apiary.io/#reference/agents/v2agentsidstatus

File: cloudbackup/cmd/shell.py (hunk 5 above)
Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: Good catch...yeah, I'll have to add a special check for that one since I only allow inputs of 1-12 for the hour.

File: cloudbackup/client/agents.py (hunk 3 above)
Comment (BryanSD) [inline cloudbackup/client/agents.py]: We're moving off of Apiary now anyways to reduce duplication. The goal is to have an internal site with the v2 documentation before it gets published publicly, but until then you can view the GitHub repository directly: https://github.com/rackerlabs/docs-cloud-backup/blob/master/rst/dev-guide/cloud-backup-v2/api-operations/methods/get-get-an-agent's-status-v2-project-id-agents-agent-id-status.rst

Comment (BenjamenMeyer) [inline cloudbackup/client/agents.py]: @jc7998 I think I'm going to put that one in a different PR as I may need to change some design up in order to make the call to the v2 API End-Point. It's going to require some thought...

File: cloudbackup/client/agents.py (hunk 4 above)
Comment (BenjamenMeyer) [inline cloudbackup/client/agents.py]: Same here as https://github.com/rackerlabs/python-cloudbackup-sdk/pull/104#discussion_r48857762

File: cloudbackup/cmd/shell.py [hunk 6]
```diff
@@ -0,0 +1,1068 @@
...
+    argument_parser.add_argument('-lg', '--log-config', default=None, type=str, dest='logconfig', help='log configuration file')
+    argument_parser.add_argument('--use-snet', default=False, action='store_true', help='Use Service Net instead of Public Net')
+
+    arguments = argument_parser.parse_args()
+
+    # If the caller provides a log configuration then use it
+    # Otherwise we'll add our own little configuration as a default
+    # That captures stdout and outputs to .agent_unique_constraint_fix-py.log
+    if arguments.logconfig is not None:
+        logging.config.fileConfig(arguments.logconfig)
```
Comment (jc7998) [inline cloudbackup/cmd/shell.py]: This line throws the following when trying to use a log file
Traceback (most recent call last):
  File "shell.py", line 1068, in <module>
    sys.exit(main())
  File "shell.py", line 1044, in main
    logging.config.fileConfig(arguments.logconfig)
AttributeError: 'module' object has no attribute 'config'

**Comment by jc7998** [comment]
When using v2, creating a recurring backup configuration fails because the timezone provided is null.
This is the entry line from the log file:
sending: {"name": "hourly", "schedule": {"recurrence": ["RRULE:FREQ=HOURLY;INTERVAL=5"], __"time_zone": null__}, "enabled": true, "notifications": [{"on_failure": true, "destination": "jorge.chai@rackspace.com", "type": "email", "on_success": true}], "inclusions": [{"path": "/home/a.txt", "type": "file"}], "agent_id": "569eef24-6976-4510-8626-82bcb1546cea", "exclusions": [], "retention": {"days": 30}}

File: cloudbackup/cmd/shell.py [hunk 7]
```diff
@@ -0,0 +1,1073 @@
...
+
+            if data['frequency'] in ('Weekly', 'Daily'):
+                promptStartTime()
+
+            if data['frequency'] == 'Hourly':
+                data['interval'] = cloudbackup.utils.menus.promptUserInputNumber(
+                    'Hourly Interval',
+                    '',
+                    0,
+                    24,
```
Comment (jc7998) [inline cloudbackup/cmd/shell.py]: I think the values need to be 1 and 23, so
1 <= X <= 23
0 and 24 hours are not valid hourly intervals

Comment (jc7998) [inline cloudbackup/cmd/shell.py]: And/or the method cloudbackup.utils.menus.promptUserInputNumber needs to be tweaked

Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: Value 0 and 24 are correct as the test is:
if X >= 0 or X <= 24 then it fails to accept the value.
this is probably more a documentation issue than anything else....I'll put some docstrings in (which I know are mostly lacking)

Comment (jc7998) [inline cloudbackup/cmd/shell.py]: @BenjamenMeyer 
I am reading that as X greater or equal to 0, so X can be 0, but the valid range should be 1 - 23 (ends included)

File: cloudbackup/cmd/shell.py (hunk 6 above)
Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: I just missed importing ```logging.config```. Fixed.

File: cloudbackup/cmd/shell.py (hunk 7 above)
Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: See https://github.com/rackerlabs/python-cloudbackup-sdk/pull/104/files#diff-a3ad6058cb9294606f9974eda3e9fd86R113

Comment (jc7998) [inline cloudbackup/cmd/shell.py]: @BenjamenMeyer 
so, should this line be updated then?
https://github.com/rackerlabs/python-cloudbackup-sdk/pull/104/files#diff-a3ad6058cb9294606f9974eda3e9fd86R101

Comment (BenjamenMeyer) [inline cloudbackup/cmd/shell.py]: Correcting the notice (https://github.com/rackerlabs/python-cloudbackup-sdk/pull/104/files#diff-a3ad6058cb9294606f9974eda3e9fd86R101) so user expectations match the logic

**Comment by BenjamenMeyer** [comment]
@jc7998 fixed the hourly backup timezone issue, tested against v1 and v2 for regressions.

**Comment by jc7998** [comment]
Looks like the formatted output for excluded files and folders uses the information for the included files.
This is for v2
https://gist.github.com/jc7998/979b9969bca8e974500d

**Comment by jc7998** [comment]
I noticed three problems when creating a daily backup and selecting a specific time:
The hour selected gets 1 subtracted. I chose 12 and the backup was created to run at 11.

The UI does not allow one to select the value of 0 when setting the minute.
The UI allows one to pass the value of 60 for minute. This value gets passed to API and the operation fails.

**Comment by BenjamenMeyer** [comment]
@jc7998 Thanks. Thinking through some changes to address what you found...moving to prompting for 24-hour format and then converting to AM/PM format for v1 API instead of the other way around too.

**Comment by jc7998** [comment]
:+1: 
The small issues can be addressed later.

**Comment by BenjamenMeyer** [comment]
This can be part of #96
